### ISSUE TITLE
Wish: QXL/Spice video

### ISSUE THREAD

**Issue by jharrison022** [stsp/dosemu2#336]
Just a pie-in-the-sky wish - QXL/Spice video support, like QEMU.

**Comment by stsp** [comment]
What does this give for the DOS apps?

**Comment by jschwartzenberg** [comment]
It would probably make sense to implement this as a plug-in at some point. This would mainly facilitate managing dosemu instances that are running remotely through libvirt.

I don't know how much work it would be to hook up libspice. Considering there are already three different video outputs (console/X11/SDL), it's probably doable. Note that not only video should be routed through libspice then, but also audio and keyboard+mouse input.

@jharrison022 are you considering developing dosemu support for libvirt?

**Comment by bolle732** [comment]
If you only want to connect a remote DOSEMU session, one can use VNC (xvnc) which I already do. Sure, a direct VNC plugin in DOSEMU would be nicer / easier.

QXL/Spice is more power hungry, but would allow for example redirecting the sound. Maybe, a serial or parallel port could be redirected too which can be nice for printing from the remote DOSEMU to a local printer ;-)

**Comment by stsp** [comment]
Guys, please, try to express yourself clearly. :)
"It would probably make sense" is not an option.
We already talked about libvirt, and no use-cases
for dosemu were spelled. Until they are clearly
spelled, "it would probably make NO sense".

@bolle732 agree about a VNC plugin. This is in
my TODO list. In fact I already prepared renderer
for multiple rendering targets, so with some coding
efforts the one can play Mortal Kombat locally with
his friend connected via vnc to the same dosemu
session! Unfortunately I've never found the time to
code this up.

Serial port can already be redirected in many different
ways including a virtual modem connection over tcp.
Parport can be redirected too.

**Comment by stsp** [comment]
> QXL/Spice is more power hungry, but would allow for example redirecting the sound.

Yes, that sounds interesting indeed.
Much better than vnc it seems.

**Comment by stsp** [comment]
> If you only want to connect a remote DOSEMU session, one can use VNC (xvnc) which I already do.

Does it work properly with SDL these days?
Last time I tried, such combo was very broken,
and the related bugs I posted to SDL bugzilla
do not seem to be fixed.

**Comment by jschwartzenberg** [comment]
Yes, Spice is just better than VNC. Once Spice is implemented, nothing extra on the dosemu side would be necessary for libvirt support.

libvirt would be nice to manage different dosemu instances with different drives, it can automatically set up a bridge to provide network connectivity and everything will have a UI, but again that would be out of scope for dosemu itself. Someone would just need to write the necessary libvirt dosemu support.

**Comment by stsp** [comment]
> libvirt would be nice to manage different dosemu instances with different drives

That probably is 0.001% of what libvirt does.
It does the extensive configuration of the VM,
but its largely irrelevant for dosemu. Even drives
cannot be successfully managed I suppose, because
it is likely targeted for images, and we instead use
a special dir layouts and synthetic symlinks.

> it can automatically set up a bridge to provide network connectivity

And dosemu can have the network connectivity
w/o bridge at all by the use of vde/slirp solution.

> Someone would just need to write the necessary libvirt dosemu support.

For what. If it would be "just need", then maybe
someone would already do that.

**Comment by jschwartzenberg** [comment]
> Even drives cannot be successfully managed I suppose, because it is likely targeted for images, and we instead use a special dir layouts and synthetic symlinks.

You suppose wrong ;) libvirt doesn't only manage VMs but also containers with special dir layouts, such as lxc. Then you also configure paths through libvirt.

> And dosemu can have the network connectivity w/o bridge at all by the use of vde/slirp solution.

So can QEMU as you know :) All configurable through libvirt.

> For what. If it would be "just need", then maybe someone would already do that.

I was thinking of it because it may help with managing many different configurations. I suspect 
jharrison022 as well, but let's see!

**Comment by bolle732** [comment]
>> If you only want to connect a remote DOSEMU session, one can use VNC (xvnc) which I already do.

> Does it work properly with SDL these days?
Last time I tried, such combo was very broken,
and the related bugs I posted to SDL bugzilla
do not seem to be fixed.

The instance I have with that set-up is up by around 10 years. It's using X, not SDL.

Last test I did was using WSVNC to have the X window of DOSEMU in a browser tab. This works with the help of 'x11vnc' and DOSEMU2 runs happily with '-X' or '-SDL'. The only problem is if you run it all on the same machine. Then your mouse will get stolen... But I think that is not the problem of DOSEMU ;-)

**Comment by stsp** [comment]
> libvirt doesn't only manage VMs but also containers with special dir layouts, such as lxc.

Well, maybe one can use libvirt to run multiple
dosemu sessions in lxcs, but there are much
simpler ways to run dosemu sessions independently.
Now the question is, how many people even want
to run multiple dosemu sessions? Its not like running
multiple qemu sessions to partition you hardware.
Do you want to use libvirt also with dosbox, with scummvm?

> So can QEMU as you know :) All configurable through libvirt.

For qemu it makes sense to keep configurable, but
for dosemu most of the non-default configs (including
bridge instead of vde) do barely matter. We simply
don't need such an advanced configurator, because the
average dosemu user change just 1-2 config options,
if at all. And you never use dosemu for partition your
HW, which is AFAIK the main use of libvirt.

> I was thinking of it because it may help with managing many different configurations.

Yes but who needs this?

**Comment by stsp** [comment]
It seems spice is largely undocumented
and is bound to qxl. Whether it would be
simple to just implement the basic rendering
without implementing qxl, is unclear to me.

**Comment by stsp** [comment]
After reading this:
https://www.spice-space.org/static/docs/spice_for_newbies.pdf
it seems spice is exceptionally qemu-specific.
Unless someone knows the simple rendering
examples, I don't want to dive into this.

**Comment by stsp** [comment]
https://en.wikipedia.org/wiki/Comparison_of_remote_desktop_software#Features

**Comment by bolle732** [comment]
But does it run without QEMU and could be used like x11vnc ? Ok, it would not be integrated as a plugin in DOSEMU, but at least one could use the Spice client to access DOSEMU.

**Comment by stsp** [comment]
Sound can be connected by the means of pulseaudio.
As for Xspice - its entirely up to user. You can start
using it right here, right now as no special support
is needed on dosemu side. But I guess you won't
start using it.

**Comment by stsp** [comment]
I guess I'd better implement old good vnc. :)

**Comment by stsp** [comment]
https://gitlab.freedesktop.org/spice/spice
libspice-server worth 2Mb of sources.

**Comment by stsp** [comment]
> Whether it would be
> simple to just implement the basic rendering
> without implementing qxl, is unclear to me.

https://www.spice-space.org/spice-user-manual.html#spice-server
```
Spice server supports the QXL VDI interface. When libspice is used with QEMU, a specific video PCI device can be used for improving remote display performance and enhancing the graphic capabilities of the guest graphic system. This video device is called a QXL device and requires guest QXL drivers for full functionality. However, standard VGA is supported when no driver exists.
```

So yes, it seems bypassing QXL is possible.

**Comment by stsp** [comment]
https://gitlab.freedesktop.org/spice/x11spice/-/blob/master/src/spice.c
x11spice is a sample application
that captures x11 display output and forwards
it to spice server, which then forwards it
to client.
It installs listeners for display events,
keyboard and tablet (mouse) to receive
display configuration events (resize etc
I suppose) and input events.
https://gitlab.freedesktop.org/spice/x11spice/-/blob/master/src/scan.c
Here it can be seen how x11 shm images
are converted into spice drawables.
I think this gives a rough overview (and
perhaps a source of code porting), except
that the audio remoting is not here.

**Comment by stsp** [comment]
> It mention Xspice, which could maybe a good starting point. Downloads are here:
> https://www.spice-space.org/download.html#XSpice

You had to point to x11spice instead,
which is the next URL on the same page
actually. :) Now I've found it myself.

**Comment by stsp** [comment]
But I firmly believe this is a courtesy
of SDL. Why would we implement yet
another rendering and input subsystems,
is unclear to me. Maintaining X11 and SDL
subsystems in parallel is already too much
of a work, and X11 almost never updated.
So I really don't want the third full-blown
back-end that hardly anyone tests.
I asked at SDL bugzilla what do they think
about adding spice or vnc.

**Comment by jschwartzenberg** [comment]
> But I firmly believe this is a courtesy
> of SDL.

Similar with Wayland support I'd say. Running graphical DOSEMU2 in Wayland without any X11 support whatsoever would seem much more relevant now than Spice.

**Comment by stsp** [comment]
But that already works because sdl
already supports wayland natively.
